### PR TITLE
Optional Attendees TDD (Tests and Test-passing code)

### PR DESCRIPTION
**Pull request by meagancheese**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -24,86 +24,136 @@
 
 public final class FindMeetingQuery {
   
-  private Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
-  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    timeOptions.clear();
+
```
Comment (jalexanderqed) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Delete extra line?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -24,86 +24,136 @@
...
+    
+    if (everyoneTimeOptions.isEmpty() && !mandatoryAttendeeEvents.isEmpty()) {
+      return getAllPossibleTimes(mandatoryAttendeeEvents, request);
+    }
+   
+    return everyoneTimeOptions;
+  }
+  
+  private Set<String> buildAttendeeSet(Collection<String> attendees) {
+    Set<String> attendeeSet = new HashSet<>();
```
Comment (jalexanderqed) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You should be able to just do `new HashSet<String>(attendees)`, in which this case function is probably unnecessary.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -24,86 +24,136 @@
...
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    List<Event> allEvents = new ArrayList<Event>();
+    allEvents.addAll(events);
+    sortByStart(allEvents);
     
-    List<Event> eventsList = new ArrayList<Event>();
-    eventsList.addAll(events);
+    List<Event> optionalAttendeeEvents = new ArrayList<Event>();
```
Comment (jalexanderqed) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It doesn't seem like optionalAttendeeEvents actually gets used at all

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -24,86 +24,136 @@
...
+      if (optionalAttendeesOnly(event, mandatoryRequestAttendees)) {
+        optionalAttendeeEvents.add(event);
+      } else {
+        mandatoryAttendeeEvents.add(event);
+      }
+    }
+    sortByStart(optionalAttendeeEvents);
+    sortByStart(mandatoryAttendeeEvents);
+  
+    Collection<TimeRange> everyoneTimeOptions = getAllPossibleTimes(allEvents, request);
```
Comment (jalexanderqed) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Wouldn't it be more efficient to build this first and return it if it's not empty, which would avoid having to build mandatoryAttendeeEvents?

**Review by jalexanderqed — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (meagancheese) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (meagancheese) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (meagancheese) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (meagancheese) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

**Comment by meagancheese** [comment]
Comments addressed, PTAL.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -24,86 +24,129 @@
...
-        return Arrays.asList(TimeRange.WHOLE_DAY);
+        attendeesWithNoEvents++;
+      }
+    }
+    if (attendeesWithNoEvents == allRequestAttendees.size()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    List<Event> allEvents = new ArrayList<Event>();
+    allEvents.addAll(events);
```
Comment (jalexanderqed) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You may be able to just do `List<Event> allEvents = new ArrayList<Event>(events);`

**Review by jalexanderqed — APPROVED** [review]
One more change (if it works), otherwise LGTM.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (meagancheese) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

**Pull request merged**